### PR TITLE
API's created for AD and LDAP users and groups

### PR DESCRIPTION
**Pull request by sonicaj**
#Ticket: 29000

### CODE REVIEW

**Comment by sonicaj** [comment]
rebuild

File: gui/api/resources.py [hunk 1]
```diff
@@ -4095,3 +4110,269 @@ def dehydrate(self, bundle):
...
+
+            del wizard_users
+
+        limit = self._meta.limit
+        if 'HTTP_X_RANGE' in request.META:
+            _range = request.META['HTTP_X_RANGE'].split('-')
+            if len(_range) > 1 and _range[1] == '':
+                limit = 0
+
+        if 'HTTP_RANGE' not in request.META:
```
Comment (william-gr) [inline gui/api/resources.py]: Shouldnt that be HTTP_X_RANGE? Why HTTP_RANGE?

File: gui/freeadmin/forms.py [hunk 2]
```diff
@@ -106,15 +107,18 @@ def _reroll(self):
...
+        kwargs = {
+            'api_name': 'v1.0',
+            'resource_name': 'account/json_users'
+        }
         if len(users) > 500:
             if self.initial:
                 self.choices = ((self.initial, self.initial),)
-            kwargs = {}
             if len(self._exclude) > 0:
                 kwargs['exclude'] = ','.join(self._exclude)
```
Comment (william-gr) [inline gui/freeadmin/forms.py]: How is the exclude being handled? It used to be a url kwarg but it doesn't seem that exist anymore and it is a GET param? So it seems like this wont work?

File: gui/api/resources.py [hunk 3]
```diff
@@ -4095,3 +4110,269 @@ def dehydrate(self, bundle):
...
+
+class JsonUserResource(DojoResource):
+
+    id = fields.CharField(attribute='id')
+    name = fields.CharField(attribute='name')
+    label = fields.CharField(attribute='label')
+
+    class Meta:
+        allowed_methods = ['get']
+        resource_name = 'account/json_users'
```
Comment (william-gr) [inline gui/api/resources.py]: Perhaps we can come up with something better for that. "account/all_users" or something?

File: gui/freeadmin/forms.py [hunk 4]
```diff
@@ -106,15 +107,18 @@ def _reroll(self):
             users = FreeNAS_Users(flags=FLAGS_DBINIT | FLAGS_CACHE_READ_USER)
         except:
             users = []
+        kwargs = {
```
Comment (sonicaj) [inline gui/freeadmin/forms.py]: @william-gr Keyword arguments are created here for the fields and the exclude key is being added to them as a string. The resource json_user accepts kwargs and checks the exclude key either in get params or in kwargs and proceeds accordingly

File: gui/api/resources.py [hunk 5]
```diff
@@ -4095,3 +4110,269 @@ def dehydrate(self, bundle):
...
+
+    def get_list(self, request, **kwargs):
+        data = {
+            'identifier': 'id',
+            'label': 'name',
+            'items': [],
+        }
+        users = []
+        query = request.GET.get('q') or kwargs.get('q')
+        exclude = request.GET.get('exclude') or kwargs.get('exclude', [])
```
Comment (sonicaj) [inline gui/api/resources.py]: @william-gr here we are getting exclude from get or kwargs

**Pull request merged**